File: homework1/Python/hsinyu_chang_task2.py
import sys
import json
from pyspark import SparkContext, SparkConf, StorageLevel
from operator import add
import time

def countPartition(iterator):
  yield sum(1 for _ in iterator)

review = sys.argv[1]
outfile = sys.argv[2]
n_partition = int(sys.argv[3])

conf = SparkConf()
conf.setMaster("local")
conf.setAppName("hw1")
sc = SparkContext.getOrCreate(conf)

reviews = sc.textFile(review)
reviewsRDD = reviews.map(lambda line: json.loads(line)).map(lambda review: (review['business_id'], 1)).persist(StorageLevel.DISK_ONLY) 

#default
default_partition = reviewsRDD.getNumPartitions()
default_n_items = reviewsRDD.mapPartitions(countPartition).collect()

defaultStart = time.time()
defaultF = reviewsRDD.reduceByKey(add).takeOrdered(10, key=lambda business: (-business[1], business[0]))
defaultEnd = time.time()
m1 = defaultEnd-defaultStart


#customized

customizedRDD = reviewsRDD.partitionBy(n_partition, lambda business: hash(business) % n_partition).persist(StorageLevel.DISK_ONLY) 
customized_n_items = customizedRDD.mapPartitions(countPartition).collect()
customizedStart = time.time()
customizedF = customizedRDD.reduceByKey(add).takeOrdered(10, key=lambda business: (-business[1], business[0]))
customizedEnd = time.time()
m2 = customizedEnd-customizedStart



task2 = {
    "default":{
        "n_partition": default_partition,
        "n_items": default_n_items,
        "exe_time": defaultEnd - defaultStart

    },
    "customized":{
        "n_partition": n_partition,
        "n_items": customized_n_items,
        "exe_time": customizedEnd-customizedStart
    },
    "explanation": "If a proper number of partitions is chosen, using a customized partitioner, hash(business_id) % n_partition, will send items with same business_id into the same bucket, therefore we can reduce the shuffling costs when we need to operate reduceBy() on the 'business_id' key and save time."
}

with open(outfile, 'w') as fd:
    json.dump(task2, fd)

File: homework1/Scala/hsinyu_chang_task2.scala
import org.apache.spark.{SparkConf, SparkContext, HashPartitioner}
import org.apache.spark.storage.StorageLevel
import org.json4s._
//import org.json4s.JsonDSL._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization.writePretty
import java.io._

object hsinyu_chang_task2 {

  implicit val formats = DefaultFormats

  def main(args: Array[String]) {

    val inputFile = args(0)
    val outputFile = args(1)
    val n_partition = args(2).toInt

    val conf = new SparkConf()

    conf.setMaster("local[*]")
    conf.setAppName("INF553")
    conf.set("spark.executor.memory", "4g")
    conf.set("spark.driver.memory", "4g")

    val sc = new SparkContext(conf)
    val reviews = sc.textFile(inputFile)

    val reviewsRDD = reviews.map(line => parse(line).extract[Map[String, String]]).
      map(review => (review("business_id"),1)).
      persist(StorageLevel.DISK_ONLY)

    val default_partition = reviewsRDD.getNumPartitions
    val default_n_items = reviewsRDD.mapPartitions(iter => Iterator(iter.size), true).collect()
    val defaultStart = System.nanoTime
    val defaultF = reviewsRDD.reduceByKey(_+_).sortBy(business => (-business._2, business._1)).take(10)
    val defaulttime = (System.nanoTime() - defaultStart)/1000000000.0


    val customizedRDD = reviewsRDD.partitionBy(new HashPartitioner(n_partition)).persist(StorageLevel.DISK_ONLY)
    val customized_n_items = customizedRDD.mapPartitions(iter => Iterator(iter.size), true).collect()

    val customizedStart = System.nanoTime
    val customizedF = customizedRDD.reduceByKey(_+_).sortBy(business => (-business._2, business._1)).take(10)
    val customizedtime = (System.nanoTime() - customizedStart)/1000000000.0



    val task2 = Map(
      "default" -> Map("n_partition" -> default_partition, "n_items" -> default_n_items, "exe_time" -> defaulttime ),
      "customized" -> Map("n_partition" -> n_partition, "n_items" -> customized_n_items, "exe_time" -> customizedtime ),
      "explanation"-> "If a proper number of partitions is chosen, using a customized partitioner, hash(business_id) % n_partition, will send items with same business_id into the same bucket, therefore we can reduce the shuffling costs when we need to operate reduceBy() on the 'business_id' key and save time.")

//
//    defaultF.foreach {
//      case (business, count) => { println(s"$business,"+f"$count%1.1f")
//      }
//    }
//
//    customizedF.foreach {
//      case (business, count) => { println(s"$business,"+f"$count%1.1f")
//      }
//    }

    val writer = new PrintWriter(new File(outputFile))
    writer.write(writePretty(task2))
    writer.close()

  }
}